### PR TITLE
build: Bump rules_go and protobuf to pick up Windows fixes

### PR DESCRIPTION
**Pull request by sesmith177**
*Description*:
- Bump rules_go to `1.15.3` (latest). This picks up some fixes so that
  rules_go works on Windows. The api go tests don't appear to build with
  `go1.11`, so pin to `go1.10.4`
- Bump `google/protobuf` to `fa252ec2`, which picks up a fix for handling
  Windows path separators

*Risk Level*:
Low
*Testing*:
Ran `bazel build //source/exe:envoy-static && bazel test //test/...` and `bazel test @envoy_api//test/... @envoy_api//tools/...`
*Docs Changes*:
n/a
*Release Notes*:
n/a

### CODE REVIEW

**Comment by sesmith177** [comment]
It looks like this commit https://github.com/bazelbuild/rules_go/commit/3a4db22f0a8e4ed6ecd9a49e4669008698884e60 breaks `bazel build @envoy_api//envoy/...`. Unfortunately, we need `rules_go 0.14.0` or later for it to have actual Windows support. Will continue looking at this

**Comment by sesmith177** [comment]
It looks like the issue is with how newer versions of `protoc-gen-go` handle a proto file importing `string.proto` (e.g. https://github.com/envoyproxy/envoy/blob/92971047ca0b8a97c1004e3b5196916203f9185b/api/envoy/type/matcher/value.proto#L7).

Specifically, it will generate a `value.pb.go` file with an invalid import statement:

```
import string "github.com/envoyproxy/data-plane-api/api/string"
```
which will fail to compile.

I've pushed a commit that works around this by renaming `string.proto` to `strings.proto`, but I'm concerned this may be considered a breaking change as the filename in import / include statements would have to change. If so, we'll have to find another way to work around this

**Comment by htuch** [comment]
@sesmith177 would you like to review/merge as is (to get this off the books) and then do a late fixup when the strings proto issue is resolved? Or just wait until then?

**Comment by sesmith177** [comment]
@htuch not sure I follow -- are you suggesting we review/merge with the renaming of `string.proto` to `strings.proto`? Or just with the `rules_go` and `google/protobuf` bumps?

**Comment by htuch** [comment]
@sesmith177 why does protoc-gen-go now do this? This seems broken.

Technically I think the string(s) rename is safe as we don't change the package namespace; see https://github.com/envoyproxy/envoy/blob/master/api/STYLE.md for discusion of when it is (un)safe to rename.

**Comment by sesmith177** [comment]
It looks like the change in behavior occurred when moving from protoc-gen-go v1.0.0 -> v1.1.0; we have not determined what caused the change.

I agree that this behavior seems like a bug

**Comment by htuch** [comment]
@sesmith177 thanks - if it's a reasonable bug to fix in protoc-gen-go, then let's do it there. Otherwise, we can probably merge this PR without breaking too much.

**Comment by sesmith177** [comment]
@htuch we submitted a PR to `golang/protobuf`: https://github.com/golang/protobuf/pull/722

One thing to note is that it is difficult to bump this dependency in Envoy, as it does not have any BUILD files -- they are generated by `rules_go` when it brings in `golang/protobuf`. So if https://github.com/golang/protobuf/pull/722 gets merged, we'll probably have to copy over some BUILD files from `rules_go` as well

**Comment by htuch** [comment]
@sesmith177 can we also bump `rules_go`?

**Comment by sesmith177** [comment]
sure, we'll bump it to `0.15.4`

**Comment by htuch** [comment]
@sesmith177 I guess I was asking "Is there an ability to fix this upstream in rules_go so we don't need the giant patch?"

**Comment by sesmith177** [comment]
@htuch it looks like if we bump rules_go forward past https://github.com/bazelbuild/rules_go/commit/fcfb5f5ac084b960a81d6bdc86d403d5f9629624, we can reference those patch files directly without needing to carry our own

File: ci/WORKSPACE.filter.example [hunk 1]
```diff
@@ -19,4 +19,4 @@ api_dependencies()
 
 load("@io_bazel_rules_go//go:def.bzl", "go_rules_dependencies", "go_register_toolchains")
 go_rules_dependencies()
-go_register_toolchains()
+go_register_toolchains(go_version = "1.10.4")
```
Comment (htuch) [inline ci/WORKSPACE.filter.example]: Do we need to modify WORKSPACE here, is there any way to hide this under a macro in `envoy_build_system.bzl`? The reason I ask this is that any change here can be chaotic to consumer of Envoy who have their own custom WORKSPACEs. Ideally we make version details as opaque as possible.

File: bazel/repositories.bzl [hunk 2]
```diff
@@ -211,7 +211,16 @@ def _go_deps(skip_targets):
     # Keep the skip_targets check around until Istio Proxy has stopped using
     # it to exclude the Go rules.
     if "io_bazel_rules_go" not in skip_targets:
+        _repository_impl(
+            name = "com_github_golang_protobuf",
+            patches = [
+                "@io_bazel_rules_go//third_party:com_github_golang_protobuf-gazelle.patch",
```
Comment (htuch) [inline bazel/repositories.bzl]: Woohoo! Can you add a comment on why these are needed and when we can get rid of the patches?

**Comment by rgs1** [comment]
@sesmith177 @htuch hmm, i am getting these errors after this change (on Linux):

```
17:38:18 ERROR: /build/tmp/_bazel_bazel/b570b5ccd0454dc9af9f65ab1833764d/external/envoy/bazel/repositories.bzl:55:13: //external:com_github_golang_protobuf: no such attribute 'patch_args' in 'git_repository' rule
17:38:18 ERROR: error loading package '': Encountered error while reading extension file 'bazel/repositories.bzl': no such package '@envoy_api//bazel': error loading package 'external': Could not load //external package
17:38:18 ERROR: error loading package '': Encountered error while reading extension file 'bazel/repositories.bzl': no such package '@envoy_api//bazel': error loading package 'external': Could not load //external package
```

It looks that `patch_args` is only in latest bazel releases:

```
~/src/bazel (master) > git show 348225e4d25                                                                                                                                      
commit 348225e4d25b9259489c1ed66eb7eca7612cddcc
Author: Klaus Aehlig <aehlig@google.com>
Date:   Wed Jun 13 05:08:16 2018 -0700

    Skylark repositories: support additional arguments for the patch tool

    Instead of hard-coding "-p0", allow the arguments for that patch tool
    to be overridden. In particular, this supports the use case of patches
    generated with `git format-patch` which are to be read as `-p1`.

~/src/bazel (master) > git tag --contains 348225e4d25b9259489c1ed66eb7eca7612cddcc
0.16.0
0.16.1
0.17.1
0.17.2
```

whereas we the minimum bazel version for envoy is 0.15.0, no?

**Comment by rgs1** [comment]
@htuch ah fair. yeah — moved to 0.17.2 and life is good again. this:

https://github.com/envoyproxy/envoy/blob/master/ci/build_container/build_container_common.sh#L3

made me think 0.15.0 was being used.

**Comment by moderation** [comment]
@rgs1 that version number if only for `buildifier`. `buildifier` 0.17.2 has just been released and it looks like they are going to mirror Bazel releases. Will generate a PR soon to bump.

**Pull request merged**